### PR TITLE
Raise an error if batch_size cannot be inferred from current batch

### LINKED ISSUE #10458: Fix `extract_batch_size` for string values and custom batches
## Proposed refactoring or deprecation

String values and custom batches are not handled correctly while extracting batch sizes.

### Motivation and Pitch

**For string values:**
Currently when we try to infer the batch_size to be used for accumulating the metrics, for string values we do this:
https://github.com/PyTorchLightning/pytorch-lightning/blob/aad86423f731d1cbc175291825da161c71123006/pytorch_lightning/utilities/data.py#L33-L34
when torch dataset returns a string, the dataloader just convert it to a list of strings and since we iterate over the list and once reach the point where we have a string we just return its len, which is not correct. More discussions here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10349

Solution 1 (I prefer):
return None and raise the warning if multiple batch sizes are found.

Solution 2:
check all the values of the iterable and see if all are strings, if they are return the len of the list. This one might increase the time complexity a little bit.

---

**For custom batches:**
we return `1`
https://github.com/PyTorchLightning/pytorch-lightning/blob/aad86423f731d1cbc175291825da161c71123006/pytorch_lightning/utilities/data.py#L41-L42
which is super silent and highly inaccurate IMO. For this one too we should return None and raise the appropriate warning.

### Additional context

As always open to suggestion/discussion/thoughts and approvals :)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #10458

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -46,6 +46,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Moved ownership of the `PrecisionPlugin` into `TrainingTypePlugin` and updated all references ([#10570](https://github.com/PyTorchLightning/pytorch-lightning/pull/10570))
 
 
+- Raised an error if the `batch_size` cannot be inferred from current batch incase if its a string or some custom batch ([#10541](https://github.com/PyTorchLightning/pytorch-lightning/pull/10541))
```
Comment (tchaton) [inline CHANGELOG.md:49]:
```suggestion
- Raised an error if the `batch_size` cannot be inferred from the current batch if it contained a string or was a custom batch object ([#10541](https://github.com/PyTorchLightning/pytorch-lightning/pull/10541))
```

File: pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -69,6 +67,13 @@ def extract_batch_size(batch: BType) -> int:
             )
             break
 
+    if batch_size is None:
+        raise MisconfigurationException(
+            "Trying to infer `batch_size` from an ambiguous collection but couldn't find one. Either"
+            " set `self.log(..., batch_size=batch_size)` or simplify the batch collection passed"
+            " inside `*_step` functions."
+        )
```
Comment (tchaton) [inline pytorch_lightning/utilities/data.py:75]:
```suggestion
        raise MisconfigurationException(
            "Trying to infer `batch_size` from an ambiguous collection but couldn't find one. Either"
            " set `self.log(..., batch_size=batch_size)`.
        )
```

Comment (tchaton) [reply pytorch_lightning/utilities/data.py:75]: Let's not introduce a new API here. It should be an int.

File: tests/utilities/test_data.py [hunk 3]
```diff
@@ -55,6 +58,20 @@ def _check_warning_raised(data, expected):
...
     _check_warning_raised(batch, 10)
 
+    # Error raised
+    batch = "test string"
+    _check_error_raised(batch)
+
+    data = {"test": ["some text"] * 7}
+    _check_error_raised(data)
+
+    class CustomBatchCollector:
```
Comment (tchaton) [inline tests/utilities/test_data.py:68]: What is this CustomBatchCollector has a len ?

File: pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/utilities/data.py:75]: We could use `-1` for that if you want to stick with ints? Although I think `None` to be more intuitive

**Review by justusschock — APPROVED** [review]

File: tests/utilities/test_data.py [hunk 4]
```diff
@@ -55,6 +58,20 @@ def _check_warning_raised(data, expected):
...
+        def __init__(self):
+            self.x = torch.randn(7, 2)
+
+    data = CustomBatchCollector()
```
Comment (rohitgr7) [inline tests/utilities/test_data.py:72]:
```suggestion
    class CustomBatch:
        def __init__(self):
            self.x = torch.randn(7, 2)

    data = CustomBatch()
```

File: tests/utilities/test_data.py (hunk 3 above)
Comment (rohitgr7) [reply tests/utilities/test_data.py:68]: It's just a custom batch class.I'll update the name

Comment (tchaton) [reply tests/utilities/test_data.py:68]: I meant, a `CustomBatch` object could have a `__len__` defined as the `batch_size`. Should we consider this too ?

File: pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/data.py:75]: Hey @rohitgr7,

As you said, this would raise an Exception. I just wanted to share that I don't understand what you meant to say here

>  Or simplify the batch collection passed " inside `*_step` functions.

To me, if the batch_size can't be inferred properly, then the user should pass the `batch_size`, not the batch.

**Review by tchaton — APPROVED** [review]
LGTM ! Waiting for open questions.

File: pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py]: there are two options. Either they can pass the batch_size to self.log(), or they can simplify the batch being passed to step functions. I have added both suggestions in the error msg.

File: tests/utilities/test_data.py (hunk 3 above)
Comment (rohitgr7) [reply tests/utilities/test_data.py]: not sure if we should do this. I haven't seen any case where user have a custom batch structure and have defined `__len__`. Plus we have `__len__` inside PT dataset so it might confuse users more.

File: pytorch_lightning/utilities/data.py [hunk 5]
```diff
@@ -58,18 +56,29 @@ def extract_batch_size(batch: BType) -> int:
     Returns:
         ``len(tensor)`` when found, or ``1`` when it hits an empty or non iterable.
     """
-    batch_size = None
-    for bs in _extract_batch_size(batch):
+    error_msg = (
+        "Trying to infer `batch_size` from an ambiguous collection but couldn't find one. Either"
+        " set `self.log(..., batch_size=batch_size)` or simplify the batch collection passed"
+        " inside `*_step` functions."
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:62]: When would the user "simplify the batch collection passed..."?

If the user has that collection structure, it's probably for a reason

File: pytorch_lightning/utilities/data.py [hunk 6]
```diff
@@ -58,18 +56,29 @@ def extract_batch_size(batch: BType) -> int:
...
+    )
+    try:
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:64]: Minimize code under the try/except

```python
batch_size = None
try:
    for bs in _extract_batch_size(batch):
        if batch_size is None:
            batch_size = bs
        elif batch_size != bs:
            warning_cache.warn(
                "Trying to infer the `batch_size` from an ambiguous collection. The batch size we"
                f" found is {batch_size}. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`."
            )
            break
except RecursionError:
    raise RecursionError(error_msg)

if batch_size is None:
    raise MisconfigurationException(error_msg)

return batch_size
```

File: pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/data.py]: Oh I understand what you meant now.

> `simplify the batch collection passed` is quite confusing.
Maybe something like. Lightning couldn't infer the `batch_size` from your batch. Either simplify its structure or provide the batch_size as self.log(..., batch_size=batch_size) `

File: tests/utilities/test_data.py (hunk 3 above)
Comment (tchaton) [reply tests/utilities/test_data.py]: I have seen several actually.

File: pytorch_lightning/utilities/data.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py]: better! updated

File: pytorch_lightning/utilities/data.py (hunk 6 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py:63]: done!

File: pytorch_lightning/utilities/data.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/utilities/data.py]: sometimes they are not aware of how the collate_fn works so I guess it's good to have.

File: pytorch_lightning/utilities/data.py [hunk 7]
```diff
@@ -58,16 +56,26 @@ def extract_batch_size(batch: BType) -> int:
...
-                "Trying to infer the `batch_size` from an ambiguous collection. The batch size we"
-                f" found is {batch_size}. To avoid any miscalculations, use `self.log(..., batch_size=batch_size)`."
-            )
-            break
+    error_msg = (
+        "We could not infer the batch_size from the batch. Either simplify its structure"
+        " or provide the batch_size as `self.log(..., batch_size=batch_size)`."
+    )
+    try:
+        batch_size = None
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:64]:
```suggestion
    batch_size = None
    try:
```

**Review by carmocca — APPROVED** [review]

**Comment by ninginthecloud** [comment]
Hi, @rohitgr7 this change breaks user's normal training flow. As auto inferencing of batch_size is very fragile, this misconfigurationexception will break down the whole training. Why not just keeping the default value?

**Comment by rohitgr7** [comment]
@ninginthecloud We have a strict rule too to only auto inference the batch_size when it's required for logging. Keeping the default value of 1 might not end up generating correct aggregated metrics.

**Comment by ninginthecloud** [comment]
> @ninginthecloud We have a strict rule too to only auto inference the batch_size when it's required for logging. Keeping the default value of 1 might not end up generating correct aggregated metrics.

Thanks for quick reply~ Your statement emphasized the importance of correctly inferencing batch_size. if inferenced batch_size is wrong, it also ends up generating wrong aggregated metrics. Therefore, we need gracefully address the limitation of auto-inferencing for batch_size. If `extract_batch_size` can't extract valid or correct value, let user know the limitation instead of throwing out exception.  I will forward the discussion to the open issue.
